### PR TITLE
DL: get match, scoring and leaderboard improvements

### PR DESCRIPTION
**Pull request by kozer**
TBA

### CODE REVIEW

File: src/GameLeague.sol [hunk 1]
```diff
@@ -5,6 +5,7 @@ import "openzeppelin-contracts/contracts/token/ERC721/utils/ERC721Holder.sol";
 import "openzeppelin-contracts/contracts/utils/Counters.sol";
 import "./CosmoShips.sol";
 import "./IRandomNumberGenerator.sol";
+import "forge-std/console.sol";
```
Comment (nikoumba) [inline src/GameLeague.sol]: No console on main contracts

File: src/GameLeague.sol [hunk 2]
```diff
@@ -173,13 +186,16 @@ contract GameLeague is ERC721Holder {
     function endEnrollmentAndStartBetting() external {
         uint256 leagueId = currentLeagueId;
         League storage league = leagues[leagueId];
+        uint256 numTeams = league.enrolledTeams.length;
+        require(numTeams >= 2 && (numTeams & (numTeams % 2)) == 0, "Number of teams must be a power of 2.");
         require(league.state == LeagueState.Initiated, "League is not in enrollment state");
         league.state = LeagueState.BetsOpen;
     }
 
     function placeBet(uint256 leagueId, uint256 teamId, uint256 amount) external {
```
Comment (nikoumba) [inline src/GameLeague.sol]: placeBet should be payable.

File: test/GameLeague.t.sol [hunk 3]
```diff
@@ -27,10 +46,63 @@ contract GameLeagueTest is Test {
...
+            cosmoShips.mint{value: mintPrice}(attributes[i], proof);
+            vm.stopPrank();
+            ids[i] = currentTokenId;
+            assertEq(cosmoShips.ownerOf(currentTokenId), recipient, "NFT not minted to correct address");
+        }
+        return ids;
+    }
+
     function testCreateTeam(string memory teamName, bool approveAll) public {
+        console.log("Running testCreateTeam");
```
Comment (nikoumba) [inline test/GameLeague.t.sol]: remove log

File: test/GameLeague.t.sol [hunk 4]
```diff
@@ -74,64 +146,41 @@ contract GameLeagueTest is Test {
     }
 
     function testInitializeLeague(uint256 _prizePool) public {
-        _prizePool = bound(_prizePool, 10 ^ 18, 10 ^ 23);
-        vm.deal(deployer, _prizePool + 10 ether);
+        console.log("Running testInitializeLeague");
```
Comment (nikoumba) [inline test/GameLeague.t.sol]: remove log

File: test/GameLeague.t.sol [hunk 5]
```diff
@@ -146,40 +195,25 @@ contract GameLeagueTest is Test {
     }
 
     function testBetPlacing() public {
+        console.log("Running testBetPlacing");
```
Comment (nikoumba) [inline test/GameLeague.t.sol]: remove log

File: test/GameLeague.t.sol [hunk 6]
```diff
@@ -146,40 +195,25 @@ contract GameLeagueTest is Test {
...
+        setupTeamAndEnroll(jane, janeAttrs, "Team-Jane");
 
         gameLeague.endEnrollmentAndStartBetting();
 
         // Alice places a bet
         uint256 betAmountAlice = 1 ether;
         vm.deal(alice, betAmountAlice);
         vm.startPrank(alice);
-        gameLeague.placeBet(leagueId, teamId, betAmountAlice);
+        gameLeague.placeBet(gameLeague.currentLeagueId(), teamId, betAmountAlice);
```
Comment (nikoumba) [inline test/GameLeague.t.sol]: see `placeBet should be payable` comment

File: test/GameLeague.t.sol [hunk 7]
```diff
@@ -225,62 +263,79 @@ contract GameLeagueTest is Test {
     }
 
     function testMatchSetupAndOutcome() public {
+        console.log("Running testMatchSetupAndOutcome");
```
Comment (nikoumba) [inline test/GameLeague.t.sol]: remove log

File: src/GameLeague.sol [hunk 8]
```diff
@@ -272,38 +291,49 @@ contract GameLeague is ERC721Holder {
...
         }
 
-        game.winner = team1Score > team2Score ? game.team1 : game.team2;
+        if (game.team1Score > game.team2Score) {
+            game.winner = game.team1;
+        } else if (game.team2Score > game.team1Score) {
+            game.winner = game.team2;
+        } else {
+            // In case of a tie, the team with the lower ID wins
+            game.winner = game.team1 < game.team2 ? game.team1 : game.team2;
```
Comment (nikoumba) [inline src/GameLeague.sol]: Isn't it better to add here also the randomness instead of awarding the older team?

File: src/GameLeague.sol [hunk 9]
```diff
@@ -314,63 +344,156 @@ contract GameLeague is ERC721Holder {
         League storage league = leagues[leagueId];
         require(league.state == LeagueState.Running, "League is not in running state");
 
-        // Run the game until only 4 teams remain
-        while (league.enrolledTeams.length > 4) {
-            // Setup matches for this turn
+        while (league.enrolledTeams.length > 1) {
```
Comment (nikoumba) [inline src/GameLeague.sol:347]: Let's put an issue here to optimize it before mainnet

File: src/GameLeague.sol [hunk 10]
```diff
@@ -314,63 +344,156 @@ contract GameLeague is ERC721Holder {
...
+        // Create a new array for remaining teams
+        uint256[] memory remainingTeams = new uint256[](league.enrolledTeams.length);
+        uint256 remainingTeamCount = 0;
 
-        // Emit leaderboard
-        uint256[] memory leaderboard;
-        for (uint256 i = 0; i < numTeams; i++) {
-            leaderboard[i] = league.enrolledTeams[i];
+        // Keep winners and eliminate losers
+        for (uint256 i = 0; i < league.enrolledTeams.length; i++) {
```
Comment (nikoumba) [inline src/GameLeague.sol:383]: Let's put an issue here to optimize it before mainnet

File: test/GameLeague.t.sol [hunk 11]
```diff
@@ -7,6 +7,7 @@ import "../src/GameLeague.sol";
 import "../src/IAttributeVerifier.sol";
 import "./fixtures/mockVerifier.sol";
 import "./fixtures/mockRandomGenerator.sol";
+import "forge-std/console.sol";
```
Comment (nikoumba) [inline test/GameLeague.t.sol]: No log

File: src/GameLeague.sol (hunk 9 above)
Comment (kozer) [reply src/GameLeague.sol:347]: https://github.com/Dymension-League/contracts/issues/4

File: src/GameLeague.sol (hunk 10 above)
Comment (kozer) [reply src/GameLeague.sol:383]: https://github.com/Dymension-League/contracts/issues/4

**Review by nikoumba — APPROVED** [review]

**Pull request merged**